### PR TITLE
Python Language

### PR DESCRIPTION
**Pull request by marikamurphy**
vote_string.py python exercise

### CODE REVIEW

**Comment by timdiller** [comment]
Please add the solutions to the exercises you have completed through Python Language.  The numpy exercise solutions can be added in another branch.

**Review by timdiller — CHANGES_REQUESTED** [review]
Please add solution files.

File: python_language/filter_words/filter_words.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+            My Bonnie lies over the sea.
+            My Bonnie lies over the ocean.
+            Oh bring back my Bonnie to me.
+         '''
+lyrics=lyrics.lower()
+lyrics=re.sub(r'[\.,]?','',lyrics)
+lyricList=lyrics.split()
+owords=[]
+for word in lyricList:
+        if(word[:1]=='o'):
```
Comment (timdiller) [inline python_language/filter_words/filter_words.py]: There is a `str` method `str.startswith()` that is the "Pythonic" way of checking if a string starts with another string.  Usually, but not always, the method of a data type is faster and more efficient than using an indexing based algorithm.  Here, the check would be `if word.startswith('o'):`

File: python_language/filter_words/filter_words.py [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+            try:
+                x=owords.index(word)
+            except:
```
Comment (timdiller) [inline python_language/filter_words/filter_words.py]: I'm always a little wary of `except` that has no specific `Exception` that it is looking for.  I _think_ in this case you are checking to see whether `word` is already in `owords`.  The error that you would want to catch here would be a `ValueError`.  When you specify `except ValueError`, then you catch that but still let other kinds of errors stop execution if they should.

Better yet, you can just do a simple `if word not in owords:` clause, which is a little more readable.

File: python_language/inventory/inventory.py [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                    frombicator      -3
+                    frombicator      20
+                    foozalator       40
+                    whitzlegidget    -4
+                    splatzleblock    -8
+                """
+                
+warehouse_log=warehouse_log.strip()
+list_log=warehouse_log.split('\n')
+log={}
```
Comment (timdiller) [inline python_language/inventory/inventory.py]: I like to encourage use of names that reflect what you're working with.  `log` here implies that it's some kind of record of transactions, when in fact, I think you are tracking inventory.  I would recommend changing the name to `inventory` or `warehouse` or similar.

File: python_language/inventory/inventory.py [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+                    splatzleblock    12
+                    frombicator      -3
+                    frombicator      20
+                    foozalator       40
+                    whitzlegidget    -4
+                    splatzleblock    -8
+                """
+                
+warehouse_log=warehouse_log.strip()
+list_log=warehouse_log.split('\n')
```
Comment (timdiller) [inline python_language/inventory/inventory.py]: "transactions" is a little more clear in indicating what the list contains.

File: python_language/inventory/inventory.py [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+log={}
+for x in list_log:
+    k,v=x.split()
```
Comment (timdiller) [inline python_language/inventory/inventory.py]: "k" is a part name, "v" is the count.  Name appropriately.

File: python_language/inventory/inventory.py [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+    try:
+        log[k]=int(v)+log[k]
+    except:
```
Comment (timdiller) [inline python_language/inventory/inventory.py]: see prior comment about the bare `except` statement here.

Better yet, you can test for membership using `if k in log` or using the `k.get(k, 0)` method, which allows you to specify a default value.

File: python_language/inventory/inventory.py [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+        log[k]=int(v)
+
+
+for k,v in log.items():
+    print('{} {}'.format(k,v))
```
Comment (timdiller) [inline python_language/inventory/inventory.py]: as a broadening exercise, think of how you could format this into a nice table with fixed-width columns, separators, etc.

File: python_language/roman_dictionary/roman_dictionary.py [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+   Clear out everyone from the dictionary you created in step 5.
+
+See :ref:`roman-dictionary-solution`.
+"""
+
+friends = {'julius': '100 via apian', 'cleopatra': '000 pyramid parkway'}
+romans = dict(brutus='234 via tratorium', cassius='111 aqueduct lane')
+countrymen = dict([('plebius','786 via bunius'), ('plebia', '786 via bunius')])
+
+print("Friends: "+str(list(friends.keys())))
```
Comment (timdiller) [inline python_language/roman_dictionary/roman_dictionary.py]: This works, but so does `print("Friends: {}".format(friends.keys())`, which is a little more compact.
`print("Friends: {}".format(", ".join(friends.keys()))` is a little more complicated, but prettier.

File: python_language/roman_dictionary/roman_dictionary.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+print("Their addresses: " +str(list(friends.values())))
+print("Altogether: "+str(list(friends.items())))
+del friends['julius']#rip
```
Comment (timdiller) [inline python_language/roman_dictionary/roman_dictionary.py]: This works fine, but I consider `friends.pop('julius')` to be a little more appropriate, and in keeping with Python's fun sensibility. ;-)

**Review by timdiller — COMMENTED** [review]
Looks good.  All of the code does what was asked, so that's a passing grade.  I have made comments that I would make for a coworker.  Also consider using the `pep8` command-line utility to conform to the [Python Style guide](https://www.python.org/dev/peps/pep-0008/), which we use at Enthought.  Many companies that use Python also promote the use of a standard coding style, and PEP8 is a common one.  (Once you get used to reading code styled this way, it makes reading the code faster and a little less wearing on the brain.)

```
c:path\to\your\repository> pep8 ./
```

File: python_language/roman_dictionary/roman_dictionary.py (hunk 8 above)
Comment (timdiller) [reply python_language/roman_dictionary/roman_dictionary.py]: To be clear, `list(friends.keys())` is redundant.  It is also fine, and perfectly acceptable, to not use `.format`: `print("Friends: " + str(friends.keys()))`

File: software_craftsmanship/refactor_inventory/refactor_inventory.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+# Import statements
+from glob import glob
+from io import open
+import os
+
+
+def findAdditionalFiles():
+
+    root=os.path.dirname(__file__)
+    log_files=glob(os.path.join(ro3884-==-ot,'data', '*.log'))
```
Comment (timdiller) [inline software_craftsmanship/refactor_inventory/refactor_inventory.py]: This looks like a typo or encoding glitch. `ro3884-==-ot`

File: software_craftsmanship/refactor_inventory/refactor_inventory.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+     
+                
+    with open(log_file, 'r', encoding='utf-8') as f:
+        read_log_file = f.read()    
+    for line in read_log_file.strip().splitlines(): # split string on lines
+        if len(line.split()) != 2: 
+            warning="{} is not valid".format(line)
+            print(warning)
+        else: 
+            warehouse_log_list.append([line.split()[0], int(line.split()[1])]) # append line
```
Comment (timdiller) [inline software_craftsmanship/refactor_inventory/refactor_inventory.py]: repeated call to `line.split()` could be avoided by doing the split on the line above:
```
part, count_str = line.split()
warehouse_log_list.append([part, int(count_str)])
```

File: software_craftsmanship/refactor_inventory/refactor_inventory.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+    return warehouse_log_list
+    
+          
+    
+def displayWarehouseLog(warehouse_log_dict): 
+    template = "{:^10.10s}|{:^10.10s}"
```
Comment (timdiller) [inline software_craftsmanship/refactor_inventory/refactor_inventory.py]: `10` here is what we call a "_magic number_" (item 4 in the refactor list) representing the column width.  Best practice is to define it explicitly as constant in the function, or to add it as an optional argument with default to the function signature.
```
def displayWarehouseLog(warehouse_lot_dict, column_width=10):
    template = "{:^" + str(column_width) + "s}|{:^" + str(column_width) + "s}"
    ...
```
That way, it's easy to change the column width, _and_ you know which "10"s are the same.  Otherwise, if you wanted to change the column width, you would have to change it multiple place.  There might coincidentally be some other magic number 10, and you would have to study to figure out if it's the _same_ 10.

File: software_craftsmanship/refactor_inventory/refactor_inventory.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+    
+    log_output='\n'+template.format("Part Name", "Count")+'\n'+'|'.join(['-' * 10] * 2)
+    for part_name, count in warehouse_log_dict.items():
+        log_output+="\n"+template.format(part_name, str(count)) 
+    log_output+="\n"
+
+    print(log_output)
+    
+def updateWarehouseLog():
+    warehouse_log_list = []
```
Comment (timdiller) [inline software_craftsmanship/refactor_inventory/refactor_inventory.py]: consider using "transactions" for "warehouse_log_list" and "inventory" for "warehouse_log_dict"

File: software_craftsmanship/refactor_inventory/refactor_inventory.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+                
+    with open(log_file, 'r', encoding='utf-8') as f:
+        read_log_file = f.read()    
+    for line in read_log_file.strip().splitlines(): # split string on lines
+        if len(line.split()) != 2: 
+            warning="{} is not valid".format(line)
+            print(warning)
+        else: 
+            warehouse_log_list.append([line.split()[0], int(line.split()[1])]) # append line
+    return warehouse_log_list
```
Comment (timdiller) [inline software_craftsmanship/refactor_inventory/refactor_inventory.py]: Generally speaking, if you are mutating the value of an object passed into the function, here `warehouse_log_list`, you shouldn't also return it.  It's certainly possible to do so in Python, but the convention is not to.  Also, in that case, you would want to name the function in a way that suggests the mutation will happen.  E.g. `update_transaction_list_from_file(log_file, transactions):` would be pretty clear.  `list.sort()` is an example of this kind of function/method.

Alternatively, there is a "functional"-style of programming that does _not_ mutate objects passed to it.  E.g. `generate_transactions_from_log_file(log_file):` would return a transaction list, which you could use to extend a previously existing list:
```
transactions = []
for log_file in log_files:
    transactions.extend(generate_transactions_from_log_file(log_file))
```
Some people prefer this style of programming because it always produces the same results given the same inputs.  Holy wars have been (and continue to be waged) over which style is better.  I try to stay out of them.

The bottom line, though, is that it's best not to return a mutated object.

File: software_craftsmanship/refactor_inventory/refactor_inventory.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+    template = "{:^10.10s}|{:^10.10s}"
+    
+    log_output='\n'+template.format("Part Name", "Count")+'\n'+'|'.join(['-' * 10] * 2)
+    for part_name, count in warehouse_log_dict.items():
+        log_output+="\n"+template.format(part_name, str(count)) 
+    log_output+="\n"
+
+    print(log_output)
+    
+def updateWarehouseLog():
```
Comment (timdiller) [inline software_craftsmanship/refactor_inventory/refactor_inventory.py:65]: consider separating this into two functions: `update_inventory_with_transactions` and `main`, which calls the functions to collect the log files, generate the transactions lists, update the inventory, and report the inventory.

**Review by timdiller — CHANGES_REQUESTED** [review]

**Review by timdiller — APPROVED** [review]
